3단계 - 자동차 경주

leedo1982 opened the pull request:
과제 3단계 구현해보았습니다

잘이해하고 작성했는지 모르겠네요 ;;

코드리뷰 감사드립니다

leedo1982 committed:
feat: step3 init

leedo1982 committed:
feat: todo 작성

leedo1982 committed:
feat: Car 객체 생성

- 상태 : 현재 위치
- 동작 : 전진한다

leedo1982 committed:
feat: 0~9 사이의 무작의 값을 구하는 기능 구현

leedo1982 committed:
feat: RacingCar 객체에서 전진 또는 멈춤 구현

시도횟수 확인을 위해 car 객체에 레이싱횟수 추가
전진의 여부기준을 car에게 변수로 전달하여 판단하도록 수정

leedo1982 committed:
feat: InputView, ResultView 로 UI 분리, RacingCarMain 구현

car 에 전진한 history 보관

namjackson commented on src/main/kotlin/study/step3/README.md:
```diff
@@ -0,0 +1,15 @@
+- Car 객체 생성 - 상태 : 현재 위치,동작 : 전진  
+- 0~9 사이의 무작의 값을 구하는 기능 구현  
+- RacingCar 객체에서 전진 또는 멈춤 구현
+- InputView, ResultView 로 UI 분리, RacingCarMain 구현 
```
요구사항 및 todo 정리 👍

namjackson commented on src/main/kotlin/study/step3/RacingCarMain.kt:
```diff
@@ -0,0 +1,14 @@
+fun main() {
+    val (numberCars, repeatRace) = input()
+
+    val racingCar = RacingCar(numberCars, repeatRace)
+    racingCar.race()
```
Exception 처리도 필요할걱같아요!

leedo1982 replied:
(자동차 경주시 발생하는 exception 을 처리해서, 사용자가 에러를 보지 못하게 하자는 의도가 맞으실까요?)
우선 전체에 try/catch 를 적용해서, 에러가 나더라도 사용자가 원인을 확인할 수 있게 print 시키도록 수정하였습니다

namjackson commented on src/main/kotlin/study/step3/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+
+class Car {
+    private var location: Int = 0
+    private var racingTimes: Int = 0
```
횟수를 의미하는 변수도 지정하셨군요 👍 
location, racingTimes등의 정보는 생성자에 주입받을수 있게하면 어떨까요?
객체 생성시, 커스텀된 프로퍼티로 조금 더 다양한 테스트하기 수월해질거 같아요! 
```kotlin
class Car(
    private var location: Int = 0
    private var racingTimes: Int = 0
)
```

leedo1982 replied:
Car 이 처음 생성될때, 외부에서 설정되면 안될것으로 생각하고, 내부에 정의해보았는데요
말씀해주신대로 적용해서  더 다양하게 확인해보겠습니다

namjackson commented on src/main/kotlin/study/step3/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+class Car {
+    private var location: Int = 0
+    private var racingTimes: Int = 0
+    private var history: MutableMap<Int, Int> = mutableMapOf()
```
불필요한 경우라면 불변변수로 지정해주세요! `val`

leedo1982 replied:
MutableMap 이어도 val 로 설정하여 처리가 가능하군요
이부분을 자주 오해했던거 같네요.
val 로 수정하겠습니다

namjackson commented on src/main/kotlin/study/step3/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+    private var racingTimes: Int = 0
+    private var history: MutableMap<Int, Int> = mutableMapOf()
+
+    fun advance(randomDigit: Int) {
```
Car의 이동하는 기준에 대한 로직이 분산되어있는것 같아요.
randomDigit을 생성해주는곳은 호출부에서 이루어지고,
생성된 randomDigit에 대한 처리는 car에서 해주고 있네요! 

현재 구조에 대한 이해가 없다면, 헷갈릴거 같아요!
advance에 왜 랜덤숫자가 필요할까? 랜덤숫자를 입력한대로 이동할까?  등등의 의문이 생길걱같아요!

이부분에 대해서 고민해보시는게 좋을거 같아요!

leedo1982 replied:
네 ,
전진할지 멈출지는 RacingCar에서 판단하고, car 은 전진하거나 멈추거나 하는 기능하는것으로 수정하겠습니다

namjackson commented on src/main/kotlin/study/step3/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+            location++
+        }
+        racingTimes++
+        history[racingTimes] = location
```
요구사항에 없는 
history까지 관리하도록 구현하셨네요 👍 
단순 history관리라면 list로 관리해도 되지않을까요?

leedo1982 replied:
결과출력을 한번에 처리하려면 history 가 유용한것으로 생각되어 만들어보았는데요.
list 로 고민하다가 list 의 index 가 시도횟수라는 의미를 가지고 있는 것 보다 
map 으로 "시도횟수 : 결과" 가 명확해보여 map 으로 해보았는데요~

list 로 변경해보겠습니다

namjackson commented on src/main/kotlin/study/step3/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+        history[racingTimes] = location
+    }
+
+    fun location(): Int = location
+    fun racingTimes(): Int = racingTimes
+    fun history(): Map<Int, Int> = history.toMap()
```
`val location: Int get() = _location` 구조나
`getLocation()` 등의 함수명은 어떨ㄲ까요?

leedo1982 replied:
넵, 함수명을 수정하도록 하겠습니다

namjackson commented on src/main/kotlin/study/step3/model/RacingCar.kt:
```diff
@@ -0,0 +1,23 @@
+        }
+    }
+
+    fun cars(): List<Car> = cars.toList()
```
방어적 복사 👍 👍 

마찬까지로 cars()라는 함수는 조금 어색한거같아요! 
`getCars()` 혹은 
`val cars get() = _cars.toList()`
은 어떨까요?

leedo1982 replied:
넵, 함수명 수정하겠습니다

namjackson commented on src/main/kotlin/study/step3/model/RacingCar.kt:
```diff
@@ -0,0 +1,23 @@
+    fun race() {
+        repeat(repeatNumber) {
+            cars.forEach { car: Car -> car.advance(randomDigit()) }
+        }
```
반복문이 중첩되네요! 함수로 분리하여 가독성을 개선해보면 어떨까요?

leedo1982 replied:
넵, 함수를 분리하겠습니다

namjackson commented on src/main/kotlin/study/step3/model/RacingCar.kt:
```diff
@@ -0,0 +1,23 @@
+    init {
+        require(numberCar > 0) { "참가 자동차 수는 0보다 커야 합니다" }
+
+        (1..numberCar).forEach { _ ->
+            cars.add(Car())
+        }
```
cars를 준비하는 함수를 만들어보는건 어떨까요?

leedo1982 replied:
넵, 함수를 분리하겠습니다

namjackson commented on src/main/kotlin/study/step3/view/InputView.kt:
```diff
@@ -0,0 +1,20 @@
+
+import study.step2.isInt
+
+class InputView {
```
객체생성이 불필요하다면, object로 선언하셔도 좋을거 같아요!

leedo1982 replied:
아하, 이런 경우 object 를 사용하면 되는군요

namjackson commented on src/main/kotlin/study/step3/view/InputView.kt:
```diff
@@ -0,0 +1,20 @@
+class InputView {
+
+    companion object {
+        fun input(): Pair<Int, Int> {
```
input이라는 이름은 너무 광범위한거 같아요.
조금더 구체적인 네이밍을 작성해보면 어떨까요?

leedo1982 replied:
넵, 수정하겠습니다

namjackson commented on src/main/kotlin/study/step3/view/InputView.kt:
```diff
@@ -0,0 +1,20 @@
+    companion object {
+        fun input(): Pair<Int, Int> {
+            println("자동차 대수는 몇 대인가요?")
+            val numberCars: String = readLine()!!
```
!! 사용은 지양하는 것이 좋습니다!
safe한 null처리 방법을 사용해주세요!

leedo1982 replied:
네, !! 제거하겠습니다

namjackson commented on src/main/kotlin/study/step3/view/ResultView.kt:
```diff
@@ -0,0 +1,20 @@
+        fun result(cars: List<Car>, racingTimes: Int) {
+            println("실행 결과")
+            (1..racingTimes).forEach { times ->
+                cars.forEach {
```
forEach가 중첩되었네요, 
for문이 여러번 중첩된 경우, 코드도 복잡해지고, 가독성도 좋지 않은거 같아요!
함수를 분리해보면 어떨까요?

leedo1982 replied:
넵, 함수로 분리하겠습니다

namjackson commented on src/test/kotlin/study/step3/model/CarTest.kt:
```diff
@@ -0,0 +1,30 @@
+
+    "주어진 자동차에, 4이상의 수를 주고 전진하면, 위치가 1 증가한다  " {
+        listOf(4, 5, 6, 7, 8, 9).forAll {
+            val car = Car()
+            val beforeLocation = car.location()
+            car.advance(it)
+            val afterLocation = car.location()
+
+            assertThat(afterLocation - beforeLocation).isOne()
```
테스트 코드 블락에 Given , when, then의 범위를 나눠보면, 테스트 코드 가독성이 좋아집니다!
```kotlin
            // given 
            val car = Car()
            val beforeLocation = car.location()

            // when
            car.advance(it)

            // then 
            val afterLocation = car.location()
            assertThat(afterLocation - beforeLocation).isOne()
```

leedo1982 replied:
넵,  범위 나눠보겠습니다

namjackson commented on src/test/kotlin/study/step3/model/RacingCarTest.kt:
```diff
@@ -0,0 +1,44 @@
+import io.kotest.matchers.throwable.shouldHaveMessage
+import org.assertj.core.api.Assertions.assertThat
+
+internal class RacingCarTest : StringSpec({
```
RacingCar에서 랜덤함수를 사용하기 때문에, 구체적인 race 함수에 대한 테스트작성이 모호해진거같아요.
Random 함수를 사용하는 경우, 어떻게 테스트를 하면 좋을까요?

랜덤 숫자를 제공하는 인터페이스를 생성하고, 구현체를 주입받아 사용하는 구조는 어떨까요?

leedo1982 replied:
RacingCar 에서 숫자를 제공하는 인터페이스를 주입받아 사용하고, 해당 구현체 활용하여 테스트 만들겠습니다

namjackson reviewed (changes requested):
3단계 자동차 경주 미션 잘구현해주셨네요! 👍 
몇가지 고민해보실법한 포인트들 코멘트로 남겼으니, 확인해주세요!
이전단게 피드백도 함께 적용해주시면 좋을거 같아요!
변경사항 적용후, 다시 리뷰 요청해주시면 감사하겠습니다!

leedo1982 committed:
feat: 코드리뷰 반영

Exception 처리
Car 객체에 location,racingTimes 의 생성자에 추가
불필요한 경우라면 불변변수로 지정 val(MutableMap)
RacingCar 에서 Car 에 대한 전진, 멈춤 판단
history관리라면 list로 변경
함수명 수정
객체생성이 불필요하다면, object 로 변경
이중for 문 함수분리
random 숫자 함수를 인터페이스화하여, 테스트 생성

leedo1982 committed:
feat: 코드리뷰 반영(문자열계산기)

문자열을 검증하는 부분과 계산하는 곳을 분리
검증시 require 사용
find 함수 사용
"1 + 1 + 1 +" 의 경우
테스트 케이스 추가

namjackson commented on src/main/kotlin/study/step3/RacingCarMain.kt:
```diff
@@ -1,14 +1,19 @@
 
 fun main() {
-    val (numberCars, repeatRace) = input()
+    try {
```
tryCatch문을 이용하여, 에러 메시지를 표시해주셧네요 👍  
코틀린에서  Result 클래스를 리턴해주는 같은 기능을 하는 runCatching을 제공해주는데요.
이런 부분을 활용하면, 좀더 가독성과 함수형스러운 코드를 작성할수 있는거 같아요!

```kotlin
       runCatching {
            // ..
        }.onSuccess { result ->
            // ..
        }.onFailure { error ->
            // ..
        }
```

leedo1982 replied:
네, runCatching 사용해보겠습니다

namjackson commented on src/main/kotlin/study/step3/util/DigitNumber.kt:
```diff
@@ -0,0 +1,5 @@
+package study.step3.util
+
+interface DigitNumber {
+    fun getDigit(): Int
```
숫자를 만들어주는 인터페이스를 작성해주셨네요 👍 👍 
다만 DigitNumber라는 이름은 숫자를 생성해준다기보다, 자료형에 가까운 이름인거같아요 
DigitGenerator, NumberGenerator 같은 네이밍은 어떨까요?

leedo1982 replied:
네, 이름 변경하겠습니다

namjackson commented on src/main/kotlin/study/step3/util/StaticDigit.kt:
```diff
@@ -0,0 +1,5 @@
+package study.step3.util
+
+class StaticDigit(private val digit: Int) : DigitNumber {
```
테스트가 힘든 랜덤 함수를 대체할수 있는 구현체  👍 👍

namjackson commented on src/main/kotlin/study/step3/model/RacingCar.kt:
```diff
@@ -1,23 +1,42 @@
+
+    private fun processRace() {
+        cars.forEach { car: Car ->
+            if (digitNumber.getDigit() >= ADVANCE_BASE_DIGIT) {
+                car.advance()
+            } else {
+                car.stop()
+            }
```
( 개인적인 의견이니, 참고만 해주셔도 좋습니다! )
현재 구조는 RacingCar 클래스내에서 차를 이동시킬지, 멈출지에 대해 판단을 하고있는데요,
차가 움직이는 책임을 Car 클래스에 옮겨 보는건 어떨까요?

Racing클래스에서는 실제 Car가 어떤 기준으로 이동하고, 동작하는지에 대해 구체적으로 알필요는 없을거같단 생각이 드는데요.
Racing클래스의 경우는 경기를 진행하는 책임을, Car는 차가 이동하는 책임을 분리하고, 
서로 협력하게 하면 좀더 객체지향적일거 같단 생각이 들어요!

leedo1982 replied:
네, 차의 이동책임을 Car 로 옮기도록 하겠습니다

namjackson commented on src/test/kotlin/study/step2/ExpressionTest.kt:
```diff
@@ -0,0 +1,58 @@
+import io.kotest.inspectors.forAll
+import io.kotest.matchers.throwable.shouldHaveMessage
+
+class ExpressionTest : StringSpec({
```
기존에 calculator에 혼합되어 테스트 되었던, 
계산 + 문자열 검증 기능을 분리하여 검증하게 되었네요! 

검증기능의 테스트 뿐아니라, 기존  calculator의 계산 테스트의 신뢰도가 늘어난거 같아요!

namjackson reviewed (commented):
3단계 자동차 경주 미션 고생하셨습니다! 
변경사항 잘 적용해주셨네요 👍 👍 
몇가지 참고하실만한 내용 코멘트로 남겼으니, 확인해주세요! 
미션 진행을 위해 머지하겠습니다! 다음 미션도 화이팅입니다 ! 😄
